pygmt/src/grdcut.py

```
@@ -133,9 +154,27 @@ def grdcut(
     with Session() as lib:
         with (
             lib.virtualfile_in(check_kind="raster", data=grid) as vingrd,
+            lib.virtualfile_in(
+                check_kind="vector", data=polygon, required=False
+            ) as vinpoly,
             lib.virtualfile_out(kind=outkind, fname=outgrid) as voutgrd,
         ):
             aliasdict["G"] = voutgrd
+
+            if polygon is not None:
+                if isinstance(polygon, PathLike) or hasattr(
+                    polygon, "__geo_interface__"
+                ):
+                    aliasdict["F"] = str(vinpoly) + aliasdict.get("F", "")
+                else:
+                    polygon_type = type(polygon).__name__
+                    err_msg = "Invalid polygon type"
+                    raise GMTValueError(
+                        err_msg,
+                        polygon_type,
+                        "Must be a PathLike, GeoDataFrame, or Shapely geometry.",
+                    )
```

This check can be done by the `data_kind` function, e.g.,
```
if polyon is not None and data_kind(polygon) not in ("file", "geojson"):
    raise GMTTypeError(...)
```
Ideally we should raise the error as early as possible, so please put it at line 140.